examples/step-27/step-27.cc

```diff
@@ -617,118 +552,6 @@ namespace Step27
         postprocess(cycle);
       }
   }
-
-
-  // @sect4{LaplaceProblem::estimate_smoothness}
-
-  // As described in the introduction, we will need to take the maximum
-  // absolute value of fourier coefficients which correspond to $k$-vector
-  // $|{\bf k}|= const$. To filter the coefficients Table we
-  // will use the FESeries::process_coefficients() which requires a predicate
-  // to be specified. The predicate should operate on TableIndices and return
-  // a pair of <code>bool</code> and <code>unsigned int</code>. The latter
-  // is the value of the map from TableIndicies to unsigned int.  It is
-  // used to define subsets of coefficients from which we search for the one
-  // with highest absolute value, i.e. $l^\infty$-norm. The <code>bool</code>
-  // parameter defines which indices should be used in processing. In the
-  // current case we are interested in coefficients which correspond to
-  // $0 < i*i+j*j < N*N$ and $0 < i*i+j*j+k*k < N*N$ in 2D and 3D, respectively.
-  template <int dim>
-  std::pair<bool, unsigned int>
-  LaplaceProblem<dim>::predicate(const TableIndices<dim> &ind)
-  {
-    unsigned int v = 0;
-    for (unsigned int i = 0; i < dim; i++)
-      v += ind[i] * ind[i];
-    if (v > 0 && v < max_degree * max_degree)
-      return std::make_pair(true, v);
-    else
-      return std::make_pair(false, v);
-  }
-
-  // This last function of significance implements the algorithm to estimate
-  // the smoothness exponent using the algorithms explained in detail in the
-  // introduction. We will therefore only comment on those points that are of
-  // implementational importance.
-  template <int dim>
-  void
-  LaplaceProblem<dim>::estimate_smoothness(Vector<float> &smoothness_indicators)
```

Comment:
we may keep the whole thing for educational purposes, but mention that exactly the same can be done via the new functionality.

Reply:
I placed a hint in the preamble of this tutorial step, at which the whole idea is described from a mathematical point of view.
https://github.com/dealii/dealii/blob/66e16cebdaf2d3a2705454206154289da58c3441/examples/step-27/doc/intro.dox#L512

We can leave the whole implementation in `step-27`, but I would rather keep it short and demonstrate how to apply those tools offered by deal.II, so that new users won't feel intimidated. The documentation in the implementation of the functions in `SmoothnessEstimator` should be sufficient for interested users.

Reply:
I think I agree with moving the implementation out. The implementation of this functionality is not terribly interesting by itself -- it doesn't really teach how to do something you may want to do with deal.II. So it's not great material for a tutorial program. It *is* great material for a tutorial program to show functions that deal.II offers to achieve a goal!